### PR TITLE
Support for updates (first draft)

### PR DESCRIPTION
**Pull request by adlrocha**
# Goal
Add support for the storage of arbitrary data (and not only immutable data). 

*This PR:*
- Renames the `internal` directory to `store`. It is a cosmetic change (and I am happy to be convinced otherwise), but how I see it,  `HashedBlockstore` uses a storethehash `store` as backend, the same way `storetheindex` will use `storethehash` for its primary storage. It is definitely a nitpick, but I think it is good for developers to perceive storethehash as something they can use for their application, and not an ad-hoc implementation for the `HashedBlockstore`.
- `store.Put` updates the value of a key if the new value is different from the one stored. If we are trying to `Put` the same value for a key it does nothing (the expected behavior for the storage of immutable data).
- Adds a `FreeList` to track primary storage slots that can be reclaimed. This is a step towards the next feature to be added: primary storage compaction.
- Minor test coverage for new changes (we will need to add way more tests in the future).

Additional context/notes: https://hackmd.io/@6vXvasZ_Q4y2vgsH5wG0uA/Sy9FbF_nu

### CODE REVIEW

File: store/freelist.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import (
+	"bufio"
+	"encoding/binary"
+	"os"
+	"sync"
+)
+
+const CIDSizePrefix = 4
+
+// A primary storage that is CID aware.
```
Comment (hannahhoward) [inline store/freelist.go]: suggestion to start thinking about breaking up this package (store package) -- it's getting quite large

File: store/index.go [hunk 2]
```diff
@@ -317,6 +317,62 @@ func (i *Index) Put(key []byte, location Block) error {
 	return nil
 }
 
+// Update a key together with a file offset into the index.
```
Comment (hannahhoward) [inline store/index.go]: Strong suggestion to test+test+test this code. This is way upstream code, and one thing I found writing it that it's pretty sensitive logic. Since we're effectively writing a database, we need to have robust testing for races/edge cases/whatever.

Comment (hannahhoward) [reply store/index.go]: *priority*

File: store/index.go [hunk 3]
```diff
@@ -317,6 +317,62 @@ func (i *Index) Put(key []byte, location Block) error {
...
+func (i *Index) Update(key []byte, location Block) error {
+	// TODO: We can probably deduplicate the code of the beginning of this
```
Comment (hannahhoward) [inline store/index.go]: I would also prioritize this TODO before adding a bunch of new functionality.

File: store/store.go [hunk 4]
```diff
@@ -159,30 +165,68 @@ func (s *Store) Put(key []byte, value []byte) error {
 		return err
 	}
 
-	has, err := s.Has(key)
+	// Get the key in primary storage
+	indexKey, err := s.index.Primary.IndexKey(key)
 	if err != nil {
 		return err
 	}
-	// we assume that a second put should NOT write twice
```
Comment (hannahhoward) [inline store/store.go:166]: rather than delete this code, can we make an option when you setup a store that allows you to specify whether a store supports updates? 

For a blockstore, a blockstore has no update, and I'm concerned this code changes behavior expected by HashedBlockstore.Put. When the hash is a function of the data itself, the code we're deleting here IS the correct behavior -- blocks are immutable, and you should never be writing a block with the same hash twice.

File: store/store.go [hunk 5]
```diff
@@ -159,30 +165,68 @@ func (s *Store) Put(key []byte, value []byte) error {
...
+	var storedKey []byte
+	var storedVal []byte
+	if found {
+		storedKey, storedVal, err = s.index.Primary.Get(prevOffset)
+		if err != nil {
+			return err
+		}
+	}
+
+	if bytes.Compare(key, storedKey)+bytes.Compare(value, storedVal) == 0 {
```
Comment (hannahhoward) [inline store/store.go]: This should be an && condition rather than sum so that you can skip the larger bytes.Compare if the first bytes.Compare is non zero. (Typical execution evaluates first expression in && and skips evaluating the second if the first fails. That's why people often write code like  `if a != nil && a.b == some value` -- the first check means the second check is safe to run without a nil pointer exception)

File: store/store.go [hunk 6]
```diff
@@ -159,30 +165,68 @@ func (s *Store) Put(key []byte, value []byte) error {
...
+	// If the key being set is not found, or the stored key is not equal
+	// (even if same prefix is shared @index), we put the key without updates
+	if !found || bytes.Compare(key, storedKey) != 0 {
+		if err := s.index.Put(indexKey, fileOffset); err != nil {
+			return err
+		}
+	}
+	// If the key exists and the one stored is the one we are trying
+	// to put this is an update.
+	if found && bytes.Compare(key, storedKey) == 0 {
```
Comment (hannahhoward) [inline store/store.go]: is this condition the else of the above condition?

Comment (hannahhoward) [reply store/store.go]: I believe they're inverses of each other.
let's say `!found`  is `a` and `bytes.Compare(key, storedKey) != 0` is `b`
the the first `if` is  `a || b`
and the second `if` is `!a && !b`
`!a && !b` is the same as `!(a || b)` so the conditions are inverses, and you can just put the update in the else block.

File: store/store.go [hunk 7]
```diff
@@ -159,30 +165,68 @@ func (s *Store) Put(key []byte, value []byte) error {
...
+	// Put value in primary storage first
+	fileOffset, err := s.index.Primary.Put(key, value)
 	if err != nil {
 		return err
 	}
-	if err := s.index.Put(indexKey, fileOffset); err != nil {
+
+	// If the key being set is not found, or the stored key is not equal
+	// (even if same prefix is shared @index), we put the key without updates
+	if !found || bytes.Compare(key, storedKey) != 0 {
```
Comment (hannahhoward) [inline store/store.go]: Follow-up -- the storedKey being different is not relevant here I think. The important thing is whether you're updating an entry in the index or writing a new one. I can see a potential problem here where the stored key is different but the indexKey is the same -- in this case, calling index.Put is going to fail.

File: store/store.go [hunk 8]
```diff
@@ -159,30 +165,68 @@ func (s *Store) Put(key []byte, value []byte) error {
...
+	// If the key exists and the one stored is the one we are trying
+	// to put this is an update.
+	if found && bytes.Compare(key, storedKey) == 0 {
+		if err := s.index.Update(indexKey, fileOffset); err != nil {
+			return err
+		}
+	}
+
+	// Add outdated data in primary storage to freelist
+	err = s.freelist.Put(prevOffset)
```
Comment (hannahhoward) [inline store/store.go]: this should only happen on update right? Should go in update block. If found != true, we should not be putting prevOffset (which I think is just 0 or -1 -- can remember how I wrote it) into the freeList no?

File: store/store.go (hunk 5 above)
Comment (hannahhoward) [reply store/store.go]: *blocking*

File: store/store.go (hunk 7 above)
Comment (hannahhoward) [reply store/store.go]: *blocking*

File: store/store.go (hunk 8 above)
Comment (hannahhoward) [reply store/store.go]: *blocking*

File: store/store.go (hunk 6 above)
Comment (hannahhoward) [reply store/store.go]: *non-blocking*

File: store/store.go (hunk 4 above)
Comment (hannahhoward) [reply store/store.go:166]: *priority*

File: store/index.go (hunk 3 above)
Comment (hannahhoward) [reply store/index.go]: *priority*

File: store/freelist.go (hunk 1 above)
Comment (hannahhoward) [reply store/freelist.go]: *non-blocking*

**Review by hannahhoward — CHANGES_REQUESTED** [review]
Hey! Overall this is excellent!

I do have some blocking changes and I'm only being a hard-ass here cause little logic errors at this low level code are going to cause pretty big problems down the road, and if we're writing a database we're gonna use in production, we're gonna have to be super diligent and not all that agile.

I've marked changes as:
*blocking* - we need to resolve before merge (or immediately before other work if we already moved to IPLD)
*priority* - not blocking for merge but not things we should leave as tech debt for a long time
*non-blocking* -- stylistic that I think isn't super important and we can wait till we get farther along

File: store/recordlist.go [hunk 9]
```diff
@@ -123,6 +124,26 @@ func (rl RecordList) Get(key []byte) (Block, bool) {
 	return blk, matched
 }
 
+// GetRecord returns the full record for a key in the recordList
+func (rl RecordList) GetRecord(key []byte) *Record {
+	// Several prefixes can match a `key`, we are only interested in the last one that
+	// matches, hence keep a match around until we can be sure it's the last one.
```
Comment (gammazero) [inline store/recordlist.go]: Is this comment correct, because it looks like the first match is returned?

File: store/recordlist.go [hunk 10]
```diff
@@ -123,6 +124,26 @@ func (rl RecordList) Get(key []byte) (Block, bool) {
...
+	rli := &RecordListIter{rl, 0}
+	for !rli.Done() {
+		record := rli.Next()
+		// The stored prefix of the key needs to match the requested key.
+		if bytes.HasPrefix(key, record.Key) {
+			return &record
+		} else if bytes.Compare(record.Key, key) == 1 {
+			// No keys from here on can possibly match, hence stop iterating. If we had a prefix
+			// match, return that, else return none
```
Comment (gammazero) [inline store/recordlist.go]: This comment does not appear correct since there is no saved prefix to return, it only returns nil at this point.

File: store/recordlist.go [hunk 11]
```diff
@@ -123,6 +124,26 @@ func (rl RecordList) Get(key []byte) (Block, bool) {
...
+func (rl RecordList) GetRecord(key []byte) *Record {
+	// Several prefixes can match a `key`, we are only interested in the last one that
+	// matches, hence keep a match around until we can be sure it's the last one.
+	rli := &RecordListIter{rl, 0}
+	for !rli.Done() {
+		record := rli.Next()
+		// The stored prefix of the key needs to match the requested key.
+		if bytes.HasPrefix(key, record.Key) {
+			return &record
+		} else if bytes.Compare(record.Key, key) == 1 {
```
Comment (gammazero) [inline store/recordlist.go]: unnecessary else

File: store/recordlist.go [hunk 12]
```diff
@@ -90,6 +90,7 @@ func (rl RecordList) PutKeys(keys []KeyPositionPair, start int, end int) []byte
 			// keys) to be bigger that that
 			(int(len(keys))*(KeySizeBytes+FileOffsetBytes+FileSizeBytes+32)))
 	newKeys = append(newKeys, rl[:start]...)
+	// Adding new keys to the beginning of the list.
 	for _, key := range keys {
 		newKeys = AddKeyPosition(newKeys, key)
 	}
```
Comment (gammazero) [inline store/recordlist.go]: This makes a temp copy of each element into `key`.  Index instead:
```go
	for i := range keys {
		newKeys = AddKeyPosition(newKeys, keys[i])
	}
```

File: store/store_test.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+func initStore(t *testing.T) (*store.Store, error) {
+	tempDir, err := ioutil.TempDir("", "sth")
+	require.NoError(t, err)
+	indexPath := filepath.Join(tempDir, "storethehash.index")
+	dataPath := filepath.Join(tempDir, "storethehash.data")
+	primary, err := cidprimary.OpenCIDPrimary(dataPath)
+	if err != nil {
+		return nil, err
+	}
+	store, err := store.OpenStore(indexPath, primary, defaultIndexSizeBits, defaultSyncInterval, defaultBurstRate)
```
Comment (gammazero) [inline store/store_test.go]: Simplify `store, err :=` to `return`

File: store/store.go (hunk 7 above)
Comment (gammazero) [inline store/store.go]: Since the result of these `bytes.Compare` calls is used later in the code, save the result in a variable so that the `bytes.Compare` does not need to be done again.

File: store/recordlist.go [hunk 14]
```diff
@@ -90,6 +90,7 @@ func (rl RecordList) PutKeys(keys []KeyPositionPair, start int, end int) []byte
 			// keys) to be bigger that that
 			(int(len(keys))*(KeySizeBytes+FileOffsetBytes+FileSizeBytes+32)))
```
Comment (gammazero) [inline store/recordlist.go]: Cast to `int` can be removed here and above

File: store/recordlist.go (hunk 12 above)
Comment (adlrocha) [reply store/recordlist.go]: That's a good one! Thanks!

File: store/recordlist.go (hunk 9 above)
Comment (adlrocha) [reply store/recordlist.go]: The implementation was wrong. I changed the comment but forgot to fix the implementation. It should be fixed (and make sense) now. Thanks!

File: store/store.go (hunk 6 above)
Comment (adlrocha) [reply store/store.go]: You are right :) Good call!

File: store/store.go (hunk 4 above)
Comment (adlrocha) [reply store/store.go:166]: We cover this case in Put. Instead of doing a `Has` here we `Get` and check the key and the value, if they are both equal we do nothing (which is the case of the Blockstore), so no behavior is changed. See: https://github.com/adlrocha/go-storethehash/blob/6798e4fc45299553187ffb75f141f2f1b1c33d9a/store/store.go#L189 (I updated this code and now we return the right error, I'll push it in a moment).

However, if you feel it is worth having an option to enable storethehash as an immutable store (as a sanity-check), we can also do that.

File: store/store.go (hunk 5 above)
Comment (adlrocha) [reply store/store.go]: That is the result of me trusting the linter. I had it like that but I changed it 🤦‍♂️. Thanks!

File: store/store.go (hunk 8 above)
Comment (adlrocha) [reply store/store.go]: Wow... good catch! I completely missed it. We definitely need some tests to catch this kind of things. Thanks!

File: store/store.go (hunk 7 above)
Comment (adlrocha) [reply store/store.go]: Another good catch! I thought I was "sanity-checking" this but I was wrong. Should be fixed by now. Thanks!

**Comment by adlrocha** [comment]
@hannahhoward, @gammazero thank you for such a thorough and quick review. I think I've addressed all your comments, the only thing missing is @hannahhoward's suggestion to test+test+test which I completely agree with. I will try to find some time to jump into that tomorrow.

However, if there is nothing blocking left in this PR, what about merging this, switching to the IPLD repo, and then having a small PR for the tests there? (that way we can start having everything in the same repo).

Thanks!

File: store/store.go (hunk 4 above)
Comment (hannahhoward) [reply store/store.go:166]: I think we should have the option for storethehash as an immutable store -- since blockstore is a big future use case. hopefully that doesn't become a burden to maintain. maybe it becomes a fork.

**Review by hannahhoward — APPROVED** [review]
